Mathlib/Data/Option/Basic.lean

```diff
@@ -3,28 +3,366 @@ Copyright (c) 2017 Mario Carneiro. All rights reserved.
+theorem pmap_map (g : γ → α) (x : Option γ) (H) :
+    pmap f (x.map g) H = pmap (fun a h => f (g a) h) x fun a h => H _ (mem_map_of_mem _ h) := by
+  cases x <;> simp only [map_none', map_some', pmap]
+
+theorem map_pmap (g : β → γ) (f : ∀ a, p a → β) (x H) :
+    Option.map g (pmap f x H) = pmap (fun a h => g (f a h)) x H :=
+  by cases x <;> simp only [map_none', map_some', pmap]
+
+-- Can't simp tag this anymore because `pmap` simplifies
+-- @[simp]
+theorem pmap_eq_map (p : α → Prop) (f : α → β) (x H) :
+    @pmap _ _ p (fun a _ => f a) x H = Option.map f x := by
+  cases x <;> simp only [map_none', map_some', pmap]
+
+theorem pmap_bind {α β γ} {x : Option α} {g : α → Option β} {p : β → Prop} {f : ∀ b, p b → γ} (H)
+    (H' : ∀ (a : α), ∀ b ∈ g a, b ∈ x >>= g) :
+    pmap f (x >>= g) H = x >>= fun a => pmap f (g a) fun b h => H _ (H' a _ h) := by
+  cases x <;> simp only [pmap, bind_eq_bind, none_bind, some_bind]
+
+theorem bind_pmap {α β γ} {p : α → Prop} (f : ∀ a, p a → β) (x : Option α) (g : β → Option γ) (H) :
+    pmap f x H >>= g = x.pbind fun a h => g (f a (H _ h)) := by
+  cases x <;> simp only [pmap, bind_eq_bind, none_bind, some_bind, pbind]
+
+variable {f x}
+
+theorem pbind_eq_none {f : ∀ a : α, a ∈ x → Option β}
+    (h' : ∀ a (H : a ∈ x), f a H = none → x = none) : x.pbind f = none ↔ x = none := by
+  cases x
+  · simp
+  · simp only [pbind, iff_false]
+    intro h
+    cases h' _ rfl h
+
+theorem pbind_eq_some {f : ∀ a : α, a ∈ x → Option β} {y : β} :
+    x.pbind f = some y ↔ ∃ (z : α) (H : z ∈ x), f z H = some y := by
+  rcases x with (_|x)
+  · simp only [pbind, false_iff, not_exists]
+    intro z h
+    simp at h
+  · simp only [pbind]
+    refine ⟨λ h => ⟨x, rfl, h⟩, ?_⟩
+    rintro ⟨z, H, hz⟩
+    simp only [mem_def, Option.some_inj] at H
+    simpa [H] using hz
+
+-- Can't simp tag this anymore because `pmap` simplifies
+-- @[simp]
+theorem pmap_eq_none_iff {h} : pmap f x h = none ↔ x = none := by cases x <;> simp
+
+-- Can't simp tag this anymore because `pmap` simplifies
+-- @[simp]
+theorem pmap_eq_some_iff {hf} {y : β} :
+    pmap f x hf = some y ↔ ∃ (a : α) (H : x = some a), f a (hf a H) = y := by
+  rcases x with (_|x)
+  · simp only [not_mem_none, exists_false, pmap, not_false_iff, exists_prop_of_false]
+  · constructor
+    · intro h
+      simp only [pmap, Option.some_inj] at h
+      refine ⟨x, rfl, h⟩
+    · rintro ⟨a, H, rfl⟩
+      simp only [mem_def, Option.some_inj] at H
+      simp only [H, pmap]
+
+-- Can't simp tag this anymore because `join` and `pmap` simplify
+-- @[simp]
+theorem join_pmap_eq_pmap_join {f : ∀ a, p a → β} {x : Option (Option α)} (H) :
+    (pmap (pmap f) x H).join = pmap f x.join fun a h => H (some a) (mem_of_mem_join h) _ rfl := by
+  rcases x with (_ | _ | x) <;> simp
+
+end Pmap
+
+@[simp]
+theorem seq_some {α β} {a : α} {f : α → β} : some f <*> some a = some (f a) :=
+  rfl
+
+@[simp]
+theorem some_orelse' (a : α) (x : Option α) : (some a).orElse (fun _ => x) = some a :=
+  rfl
```

Comment:
I guess we want lots of renames here. All the `orelse` fragments should become `orElse`, similarly for `isSome` and `isNone`, and then all the necessary #aligns.